network/ieee802_11_prism.ksy

```diff
@@ -0,0 +1,113 @@
+meta:
+  id: ieee802_11_prism
+  title: prism II headers for 802.11 capture
+  license: Unlicense
+  endian: le
+  encoding: ascii
+  xref:
+    wikidata: Q7245672
+doc: |
+  Prism II headers are usually used by wlan cards based on Intersil / Conexant [Prism II](https://en.wikipedia.org/wiki/Prism_(chipset)).
+
+doc-ref: https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html
+
+seq:
+  - id: message_code
+    type: u4
+  - id: size
+    type: u4
+  - id: device_name
+    type: strz
+    size: 16
+  - id: fields
+    type: fields
+    size: fields_size
+  - id: body
+    size-eos: true
+instances:
+  header_size:
+    value: 4 + 4 + 16
+  fields_size:
+    value: size - header_size
+types:
+  fields:
+    seq:
+      - id: fields
+        type: field
+        repeat: eos
+    types:
+      field:
+        seq:
+          - id: header
+            type: header
+          - id: is_not_supplied
+            type: u2
+          - id: size
+            type: u2
+          - id: value
+            size: size
+            type:
+              switch-on: header.type
+              cases:
+                'type::host_time': u4
+                'type::mac_time': u4
+                'type::channel': u4
+                'type::rssi': u4
+                'type::signal_quality': u4
+                'type::noise': u4
+                'type::data_rate': u4
+                'type::is_tx': u4
+                'type::frame_length': u4
```

Comment:
This page http://www.martin.cc/linux/prism actually has some interesting information about the representation of these values, could you check it out and possibly update your spec?

Reply:
Thanks, in fact I have looked into it, but it seemed like it just a duplicate of the info from the source from doc-ref.

I will look into it again more carefully.

Reply:
Could you point out what in your opinion should be added?

Reply:
> Could you point out what in your opinion should be added?

1. 
   ```
    * rssi:     Appears to be a signed dbm value
   ```
   
   ```
    * noise:     (...) - Should be signed dbm value
   ```
   
   These values should be **signed (`s4`)**, not ~~unsigned (`u4`)~~ as in your spec.

2.
   ```
    * rate:     Appears to be in units/multiples of 500Khz
   ```
   
   Maybe it would make sense to declare a value instance to get the value in standard units like `kHz` or `Hz` .
   
   Additionally, using [verbose enums](https://doc.kaitai.io/user_guide.html#verbose-enums) you document the    `type::data_rate` enum value like this:
   
   ```yaml
   enums:
     type:
       # ...
       0x8:
         id: data_rate
         doc: in units/multiples of 500Khz
   ```

   Or maybe reflect the unit in the name of the enum value, like `data_rate_div_500khz` (meaning that the value stored here is calculated like `rate in Hz (basic unit) / 500 kHz`) or something like that.

3.
   ```
    * istx:     Not Supplied  - 0 = rx packet, 1 = tx packet
   ```
   
   From a raw `'type::is_tx': u4` type, this value meaning is not known to the end user. Best practice in terms of clarity and usability would be to create a value instance returning an enum value of something like `enums/packet_direction: {0: rx, 1: tx}`, but merely documenting this value logic in the `type::is_tx` enum value is also OK.

The info about how to interpret the `u4` values in context of individual field types is essential and thus should IMHO definitely appear somewhere in the KSY spec. At the very least, as a YAML comment `# ...` next to the `type/cases` items, a better option would be `doc` key in [verbose enums](https://doc.kaitai.io/user_guide.html#verbose-enums) (because these docblocks could then appear in auto-generated API documentations and autocompletion boxes when typing the application code in a smart IDE), and the best option is the code that _documents itself_, i.e. it is clear from the attribute names what's the meaning of the value in the attribute, in which units it is, etc. However, this is not always possible/convenient, in these cases it should be accompanied with the docblock.

---

My opinion is that stopping parsing on "magic" `u4` integers without any effort to try to document their meaning is insufficient, and it forces all potential users of the KSY spec to do it on their own. Of course, if it is the best we can do with regard to publicly available resources, then it is understandable. But if we can make the KSY spec better and more usable with minimal effort, IMHO we should always do it.

Reply:
Gonna do.

BTW: https://github.com/kaitai-io/kaitai_struct/issues/522 .